[WebRTC] Enable assembly optimizations in BoringSSL build

mstyura opened the pull request:
#### 2628ca02ea1f71c1cfad3290944349e35d4450bc
<pre>
[WebRTC] Enable assembly optimizations in BoringSSL build

<a href="https://bugs.webkit.org/show_bug.cgi?id=295573">https://bugs.webkit.org/show_bug.cgi?id=295573</a>

Reviewed by Youenn Fablet.

Include assembly files from BoringSSL into compilation to improve
performance of RTP encryption and decryption.
Defined OPENSSL_SMALL during compilation, matching the compilation
of BoringSSL with BUILD.gn as used in Chromium.

Enabling hardware-accelerated AES changes the cipher preference in BoringSSL,
which causes all assertions in crypto-suite.https.html tests to pass.

* LayoutTests/imported/w3c/web-platform-tests/webrtc/protocol/crypto-suite.https-expected.txt:
* Source/ThirdParty/libwebrtc/Configurations/boringssl.xcconfig:
* Source/ThirdParty/libwebrtc/libwebrtc.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/299383@main">https://commits.webkit.org/299383@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1e345217cc782588ae2080e4aa88959abbf00b41

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118396 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38076 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28723 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124560 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70449 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38771 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46658 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89861 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59455 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121349 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30887 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106155 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70338 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29941 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24263 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68223 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100317 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24453 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127631 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45302 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34165 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98515 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45665 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102375 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98302 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25068 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43718 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21705 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/41800 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45172 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44635 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47979 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46322 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/ThirdParty/libwebrtc/Configurations/boringssl.xcconfig:
```diff
@@ -13,7 +13,6 @@ PUBLIC_HEADERS_FOLDER_PATH = $(INSTALL_PATH_PREFIX)$(WK_LIBRARY_INSTALL_PATH)/li
 
-// FIXME: There's no assembly used in the boringssl build yet :(
-GCC_PREPROCESSOR_DEFINITIONS = OPENSSL_NO_ASM $(inherited);
+GCC_PREPROCESSOR_DEFINITIONS = OPENSSL_SMALL $(inherited);
```
Why do we need `OPENSSL_SMALL`?

mstyura replied:
I could omit the define, but then I'll need to include [`Source/ThirdParty/libwebrtc/Source/third_party/boringssl/src/third_party/fiat/curve25519_64_adx.h`](https://github.com/WebKit/WebKit/blob/main/Source/ThirdParty/libwebrtc/Source/third_party/boringssl/src/third_party/fiat/curve25519_64_adx.h) into `xcodeproj` (required for `x86_64`).

I've noticed that [WebRTC in Google Chrome](https://chromium.googlesource.com/chromium/src/third_party/boringssl/+/refs/heads/main/BUILD.gn#37) is compiled with `OPENSSL_SMALL` and does not require `curve25519_64_adx.h` to be used, so I've decided to define `OPENSSL_SMALL` for WebKit.
Let me known what way you want to proceed taking into account provided information.

I've found the following notes about `OPENSSL_SMALL` in documentation:
[BUILDING.md](https://boringssl.googlesource.com/boringssl/+/HEAD/BUILDING.md):
> In order to serve environments where code-size is important as well as those where performance is the overriding concern, OPENSSL_SMALL can be defined to remove some code that is especially large.

[INCORPORATING.md](https://boringssl.googlesource.com/boringssl/+/HEAD/INCORPORATING.md)
> OPENSSL_SMALL removes some code that is especially large at some performance cost.

youennf replied:
Could you add this information to the commit message?
In particular that Chrome is compiling with it.

youennf commented on LayoutTests/imported/w3c/web-platform-tests/webrtc/protocol/crypto-suite.https-expected.txt:
```diff
@@ -1,8 +1,8 @@
-FAIL dtlsCipher is acceptable on data-only assert_true: expected true got false
-FAIL dtlsCipher is acceptable on video-only assert_true: expected true got false
+PASS dtlsCipher is acceptable on data-only
+PASS dtlsCipher is acceptable on video-only
```
So enabling assembly add some cipher support. Do you know why?

mstyura replied:
Yes, I've figured it out when prepared PR. The reason is not the new cipher suite is added, but the cipher suites order (preference) has changed. The order depends on the presence of HW AES support. The ordering defined here:
https://github.com/WebKit/WebKit/blob/c46918d1cba365017d1df98e3d14f04db42320bc/Source/ThirdParty/libwebrtc/Source/third_party/boringssl/src/ssl/ssl_cipher.cc#L1053-L1077

youennf replied:
Could you add this information to the commit message?

youennf reviewed (approved):
LGTM

mstyura committed:
[WebRTC] Enable assembly optimizations in BoringSSL build

https://bugs.webkit.org/show_bug.cgi?id=295573

Reviewed by Youenn Fablet.

Include assembly files from BoringSSL into compilation to improve
performance of RTP encryption and decryption.
Defined OPENSSL_SMALL during compilation, matching the compilation
of BoringSSL with BUILD.gn as used in Chromium.

Enabling hardware-accelerated AES changes the cipher preference in BoringSSL,
which causes all assertions in crypto-suite.https.html tests to pass.

* LayoutTests/imported/w3c/web-platform-tests/webrtc/protocol/crypto-suite.https-expected.txt:
* Source/ThirdParty/libwebrtc/Configurations/boringssl.xcconfig:
* Source/ThirdParty/libwebrtc/libwebrtc.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/299383@main
